### ISSUE TITLE
ctr cann't export image pulled by crictl

### ISSUE THREAD

**Issue by johnsonisjiang** [kubernetes-sigs/cri-tools#1636]
#### What happened:
when in containerd environment, if we use old version crictl and ctr, no matter the images pulled by ctr or crictl, the ctr can export the image successfully.
But when crictl >= 1.24.0, the issue happens, the image pulled by crictl, when use ctr export, it will failed;  For comparison, images pulled by ctr  have no problem.

ctr export image which pulled by crictl  , will show error message

**ctr  -n k8s.io images export my_image.tar sha256:af282666faecdad41485bea589ff4c6f25af2abbf76c7b8dff24e7df2570798e
ctr: failed to get reader: content digest sha256:10d56570098259f40562e64b4d5ad9374c94af3e00688d62827e8c79c47d9617: not found**

the crictl version: 1.25.0, ctr version 1.7.20

#### What you expected to happen:
I have another env (crictl:1.19.0; ctr:1.3.7), in this machine, images pulled by crictl can be  exported by ctr successfully.
I think there maybe some changes between crictl versions, which changes the export perform.
Would you please tell me how to export images pulled by new version crictl?

#### Environment:

the crictl version: 1.25.0, ctr version 1.7.20

**Comment by saschagrunert** [comment]
THank you for opening this report @johnsonisjiang. Both `crictl` v1.19.0 and v1.25.0 are end of life. `crictl` also just invokes the RPC to pull the image, so the main issue seems to be on the containerd side here.

**Comment by johnsonisjiang** [comment]
> THank you for opening this report @johnsonisjiang. Both `crictl` v1.19.0 and v1.25.0 are end of life. `crictl` also just invokes the RPC to pull the image, so the main issue seems to be on the containerd side here.

Thanks, I got the reason, if the "/etc/containerd/config.toml"->"discard_unpacked_layers" was true, the image pulled by crictl may exported failed, but it have no effect on images pulled by ctr(ctr use contained api not grpc with containerd, default was false)

**Comment by saschagrunert** [comment]
@johnsonisjiang unfortunately I don't think there is much we can do from a cri-tools perspective.

**Comment by saschagrunert** [comment]
I'm going to close this one now, feel free to reopen if there is any action we can take.

**Comment by SergeyKanzhelev** [comment]
@johnsonisjiang please file a bug with containerd if this is an important scenario for you. In theory the `discard_unpacked_layers` should not be affecting image exporting. Or at least this must be documented clearly.

**Comment by samuelkarp** [comment]
> In theory the `discard_unpacked_layers` should not be affecting image exporting. 

`discard_unpacked_layers` causes the original compressed tars for layers to be discarded from the content store.  Those layers have a specific digest, computed over the compressed tar.  It is not possible to reliably reproduce this compressed tar; file ordering within the tar and differences in compression (either via different settings on the same implementation or entirely different gzip implementations) will affect the resulting file and thus the digest.

In other words: this is working-as-intended if `discard_unpacked_layers` is set to `true`.  If you expect to export images from containerd, you should set `discard_unpacked_layers` to `false`.

**Comment by ErikEngerd** [comment]
In my setup (kubeadm kubernetes with containerd 1.6.20~ds1-1+b1, debian 12.9), I have discard_unpacked_layers = false. However, export is failing because layers of pulled images are not there. They are directly put into snapshots so the flag appears to be ignored. Running the images using 'ctr run' works so the image is definitely available. 

I am looking for a way to enable this so I can do exports and imports of images. Alternatively, it would also be an idea for export and import snapshots (it at all possible). The latter option would be a bit better since it would require less disk use. I am looking into a solution for pre-pulling images between kubernetes nodes without having to run a caching proxy. It is relatively easy to do using a daemonset. If can do local export and import using the go containerd api then the rest is just network communication.
